### PR TITLE
Verify that preloads with a "fetch" `as` value have an empty destination

### PR DESCRIPTION
**Pull request by yoavweiss**
@annevk - As discussed on https://github.com/whatwg/fetch/pull/442, here's the corresponding test. Note that it fails in all current implementations.

### CODE REVIEW

File: preload/fetch_destination.https.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="../service-workers/service-worker/resources/test-helpers.sub.js"></script>
+<script>
+async_test(function(t) {
+    var worker_url = 'resources/fetch-destination-worker.js';
+    var scope = 'resources/empty.html';
```
Comment (annevk) [inline preload/fetch_destination.https.html]: You can use `/common/blank.html` instead.

Comment (annevk) [inline preload/fetch_destination.https.html]: Also, that URL doesn't match the preload URL from later on. Is that intentional?

**Comment by yoavweiss** [comment]
@annevk - Changed the scope to be the preloaded resource, and added a bunch of needed step_func wrappers around the various async test functions. (due to timeouts I seen locally)

File: preload/resources/fetch-destination-worker.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+self.addEventListener('fetch', function(event) {
+    if (event.request.url.indexOf('dummy.xml') != -1) {
+        var response = new Response();
+        if (event.request.destination == "")
+            response.status = 200;
+        else
+            response.status = 500;
```
Comment (annevk) [inline preload/resources/fetch-destination-worker.js]: This doesn't match the Preload specification. It doesn't say that an error event is dispatched for such status codes. That would also likely be problematic due to the same-origin policy mostly trying to keep that a secret.

File: preload/fetch_destination.https.html [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      .then(t.step_func(function() {
+          return with_iframe(scope);
+        }))
+      .then(t.step_func(function(frame) {
+          var link = frame.contentWindow.document.createElement("link");
+          link.as = "fetch";
+          link.href = scope;
+          link.rel = "preload";
+          link.addEventListener("load", t.step_func(function() { done(registration); }));
+          link.addEventListener("error", t.step_func(function() { assert_unreached("Fetch destination preload errored"); done(registration); }));
```
Comment (annevk) [inline preload/fetch_destination.https.html]: Does this even run? I think that an assert results in an exception.

**Review by annevk — CHANGES_REQUESTED** [review]
Found some more issues.

File: preload/resources/fetch-destination-worker.js (hunk 2 above)
Comment (yoavweiss) [inline preload/resources/fetch-destination-worker.js]: Hmm, I guess we need to tighten the definition of "load was successful" there. I understood it as including the status code. Is there another way that a SW can signal the page that a load was unsuccessful?

File: preload/fetch_destination.https.html (hunk 3 above)
Comment (yoavweiss) [inline preload/fetch_destination.https.html]: It runs and fails in current implementations (which trigger an error when `as=fetch`)

File: preload/resources/fetch-destination-worker.js (hunk 2 above)
Comment (annevk) [inline preload/resources/fetch-destination-worker.js]: I'm not sure. @jakearchibald?

File: preload/fetch_destination.https.html (hunk 3 above)
Comment (annevk) [inline preload/fetch_destination.https.html]: But does it call `done()`?

Comment (yoavweiss) [inline preload/fetch_destination.https.html]: It fails (immediately) with an assert_unreached().
If this is an issue and done() must be called, what's the proper way define unreachable code without asserting?

Comment (annevk) [inline preload/fetch_destination.https.html]: I don't think it's an issue, but I would just leave out done(). Come to think of it, you can use `t.step_func_done()` which cleans this up automatically.

File: preload/resources/fetch-destination-worker.js (hunk 2 above)
Comment (annevk) [inline preload/resources/fetch-destination-worker.js]: In theory returning a network error can be done through `Response.error()` per https://fetch.spec.whatwg.org/#dom-response-error.

Comment (yoavweiss) [inline preload/resources/fetch-destination-worker.js]: Cool. Using that instead

File: preload/fetch_destination.https.html (hunk 3 above)
Comment (yoavweiss) [inline preload/fetch_destination.https.html]: `step_func_done` it is (plus unregistration)

**Comment by yoavweiss** [comment]
@annevk - I brought back the old scope (with the empty.html) as I was hitting weird errors related to frame access with the scope and link both pointing at the same URL.

**Comment by annevk** [comment]
What's wrong with using http://w3c-test.org/common/blank.html as I suggested?

**Comment by yoavweiss** [comment]
Oh, should've commented on that. It's from a different directory than the SW registered, and the SW registration fails.

File: preload/fetch_destination.https.html [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        }))
+      .then(t.step_func(function() {
+          return with_iframe(scope);
+        }))
+      .then(t.step_func(function(frame) {
+          var link = frame.contentWindow.document.createElement("link");
+          link.as = "fetch";
+          link.href = "resources/dummy.xml";
+          link.rel = "preload";
+          link.addEventListener("load", t.step_func(function() { done(registration); }));
```
Comment (annevk) [inline preload/fetch_destination.https.html]: Use step_func_done here too and inline the unregister call?

Seems okay then.

Comment (yoavweiss) [inline preload/fetch_destination.https.html]: done

**Comment by yoavweiss** [comment]
Would you like me to squash the commits?

**Pull request merged**